### ISSUE TITLE
deprecate config using the One True ADS Server(tm)

### ISSUE THREAD

**Issue by markdroth** [envoyproxy/envoy#13951]
Today, the way ADS is configured is as follows:
- The bootstrap file has an [`ads_config`](https://github.com/envoyproxy/envoy/blob/d9af0e985a2fef2c921641b12f4786a658ad0923/api/envoy/config/bootstrap/v3/bootstrap.proto#L100) field indicating the One True ADS Server(tm).
- The bootstrap file has its [`lds_resource_locator`](https://github.com/envoyproxy/envoy/blob/d9af0e985a2fef2c921641b12f4786a658ad0923/api/envoy/config/bootstrap/v3/bootstrap.proto#L82) and [`cds_resources_locator`](https://github.com/envoyproxy/envoy/blob/d9af0e985a2fef2c921641b12f4786a658ad0923/api/envoy/config/bootstrap/v3/bootstrap.proto#L91) fields set to a `ConfigSource` with the [`ads`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L185) field set.
- In resources sent from the management server, `ConfigSource` protos use the same [`ads`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L185) field.

This approach seems incompatible with the new xDS naming scheme, which was designed to support federation.  As part of the new naming scheme, the `ApiConfigSource.ApiType` enum now has [`AGGREGATED_GRPC`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L68) and [`AGGREGATED_DELTA_GRPC`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L73) values, which allows configuring any `ConfigSource` to use a single ADS stream instead of using a separate stream for each resource type.  This way of configuring ADS seems much more in-line with the direction we're heading in and will make it easier to transition to the new federated world.

Therefore, I propose replacing this notion of the One True ADS Server(tm) with use of this new `AGGREGATED_GRPC` enum value.  The typical ADS configuration described above would instead be represented as follows:
- The bootstrap file will not populate the [`ads_config`](https://github.com/envoyproxy/envoy/blob/d9af0e985a2fef2c921641b12f4786a658ad0923/api/envoy/config/bootstrap/v3/bootstrap.proto#L100) field.
- The bootstrap file will have its [`lds_resource_locator`](https://github.com/envoyproxy/envoy/blob/d9af0e985a2fef2c921641b12f4786a658ad0923/api/envoy/config/bootstrap/v3/bootstrap.proto#L82) and [`cds_resources_locator`](https://github.com/envoyproxy/envoy/blob/d9af0e985a2fef2c921641b12f4786a658ad0923/api/envoy/config/bootstrap/v3/bootstrap.proto#L91) fields both set to the same server, with `ApiType` set to `AGGREGATED_GRPC` or `AGGREGATED_DELTA_GRPC`.  (Envoy will internally de-dup these by hashing the `ConfigSource`s, so the result will be a single ADS stream, not two separate ADS streams.)
- In resources sent from the management server, `ConfigSource` protos will use the [`self`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L198) field.

This will require the following changes:
- Implement support for the [`self`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L198) field.  (This field is not currently implemented in Envoy.)
- Implement support for the [`AGGREGATED_GRPC`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L68) and [`AGGREGATED_DELTA_GRPC`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L73) values in the `ApiType` enum.  (I'm not sure how much of this works right now.)
- Deprecate the bootstrap file's [`ads_config`](https://github.com/envoyproxy/envoy/blob/d9af0e985a2fef2c921641b12f4786a658ad0923/api/envoy/config/bootstrap/v3/bootstrap.proto#L100) field.
- Deprecate the [`ads`](https://github.com/envoyproxy/envoy/blob/e1c138b427ff9bcdf5ebcb2639cc2d22f98ac1a8/api/envoy/config/core/v3/config_source.proto#L185) field.

The only configuration that would be broken by this change would be if someone was (e.g.) using non-aggregated LDS to get a `Listener` resource that contained a `ConfigSource` telling the client to get the `RouteConfiguration` via the One True ADS Server(tm).  However, it seems exceedingly unlikely that anyone is actually doing that, and there are work-arounds: they can either have their management server send the exact `ConfigSource` containing the ADS server they want to talk to, or better yet, they can start using the new xDS naming scheme, where they can select the xDS server via an authority in the resource name itself.

CC @mattklein123 @htuch

**Comment by htuch** [comment]
I agree with this directionally, suggest waiting until we have minor versioning working to execute.

**Comment by nezdolik** [comment]
@markdroth @htuch is anyone working on this? i can help out

**Comment by markdroth** [comment]
I think @adisuissa was going to look at this, but I don't know if he's started yet.

**Comment by adisuissa** [comment]
In addition to the problem with non-aggregated listeners described in the [first comment](https://github.com/envoyproxy/envoy/issues/13951#issue-739457652), I think there are other use-cases that may not fit with the proposal. Specifically, `self` limits which ADS config to use. Here are a couple of examples:
1. Using static clusters, where different sub-groups of clusters each wants to use a different ADS config source for their endpoints.
2. Using one ADS for clusters, endpoints, listeners, etc. and another ADS for secrets.

A proposal for named config-source that gets around this issue, and suggests a path forward on future xDSTP features, can be found [here](https://docs.google.com/document/d/1rIH_ScIM4gUR12ummIpxRY0r_gsxgaLowuiSAYUTSFU/edit#).

**Comment by markdroth** [comment]
@adisuissa I think both of those use-cases are actually already handled by the approach described in this issue.  I don't see how the proposal for named ConfigSources actually helps with any of this.

Note that there is no requirement here to use "self" ConfigSources for ADS.  The "self" option is useful in that it prevents the need for control planes to embed their own name in their resources, but that option really has nothing to do with ADS specifically.  The intention here is that a ConfigSource can specify any server it wants and can request the use of ADS by setting the `ApiType` enum to `AGGREGATED_GRPC`, so it should be possible to get whatever resources you want from whatever ADS streams you want.

**Comment by adisuissa** [comment]
> Note that there is no requirement here to use "self" ConfigSources for ADS. The "self" option is useful in that it prevents the need for control planes to embed their own name in their resources, but that option really has nothing to do with ADS specifically. The intention here is that a ConfigSource can specify any server it wants and can request the use of ADS by setting the `ApiType` enum to `AGGREGATED_GRPC`, so it should be possible to get whatever resources you want from whatever ADS streams you want.

My concern is the need to rewrite the same config source in many places, and then the need to identify the same source by comparing the protos (it can be hashed, but not ideal). Making the names explicit has the benefit of defining it once and referencing it from multiple places.

**Comment by markdroth** [comment]
If the goal is to avoid the repetition, then using xdstp names gives you that, where the name is the authority.  Why do we need a second mechanism to do the same thing?

**Comment by adisuissa** [comment]
I agree that xdstp naming gives that, but requires changes on the server side.
In addition the named-config-source allows decoupling the stream from the authority.
For example, if a single authority provides both secrets and other config, but wants to serve them from different servers, then the authority in the resource names can stay the same, but the providing server could be different.

**Comment by markdroth** [comment]
The named ConfigSource approach also requires server-side changes, so I don't see how that's a benefit here.

The whole point of the authority in the xdstp names is to indicate what server to get the resource from.  There's no value in having the same authority for two resources that you want to fetch from different servers.

I don't see what this approach gives us that we don't already have.  I think we should implement the thing we've already designed rather than add a second way to do the same thing.

**Comment by htuch** [comment]
xdstp:// changes aren't really the huge IMHO. You can even start by using them as opaque identifiers in server or client.
